proposals/ideas/xxx-flashflow.txt

```diff
@@ -0,0 +1,713 @@
+
+In the short term we will implement a working FlashFlow measurement system.
+This requires code changes in little-t tor and an external FlashFlow codebase.
+The majority of the implementation work will be done in the short term, and
+the product is a complete FlashFlow measurement system. Remaining pieces (e.g.
+better authentication) are added later for enhanced security and network
+performance.
+
+In the medium term we will begin collecting data with a FlashFlow
+deployment. The intermediate results and v3bw files produced will be made
+available (semi?) publicly for study.
+
+In the long term experiments will be performed to study ways of using FF v3bw
+files to improve load balancing. Two examples: (1) using FF v3bw files instead
+of sbws's (and eventually phasing out torflow/sbws), and (2) continuing to run
+sbws but use FF's results as a better estimate of relay capacity than observed
+bandwidth. Authentication and other FlashFlow features necessary to make it
+completely ready for full production deployment will be worked on during this
+long term phase.
+
+3. FlashFlow measurement system: Short term
+
+The core measurement mechanics will be implemented in little-t tor, but a
+separate codebase for the FlashFlow side of the measurement system will also
+be created. This section is divided into three parts: first a discussion of
+changes/additions that logically reside entirely within tor (essentially:
+relay-side modifications), second a discussion of the separate FlashFlow
+code that also requires some amount of tor changes (essentially: measurer-side
+and coordinator-side modifications), and third a security discussion.
+
+3.1 Little-T Tor Components
+
+The primary additions/changes that entirely reside within tor on the relay
+side:
+
+  - New torrc options/consensus parameters.
+  - New cell types.
+  - Pre-measurement handshaking (with a simplified authentication scheme).
+  - Measurement mode, during which the relay will echo traffic with measurers,
+    set a cap on the amount of background traffic it transfers, and report the
+    amount of transferred background traffic.
+
+3.1.1 Parameters
+
+FlashFlow will require some consensus parameters/torrc options. Each has some
+default value if nothing is specified; the consensus parameter overrides this
+default value; the torrc option overrides both.
+
+FFMeasurementsAllowed: A global toggle on whether or not to allow
+measurements. Even if all other settings would allow a measurement, if this is
+turned off, then no measurement is allowed. Possible values: 0, 1. Default: 0
+(disallowed).
+
+FFAllowedCoordinators: The list of coordinator TLS certificate fingerprints that are
+allowed to start measurements. Relays check their torrc when they receive a
+connection from a FlashFlow coordinator to see if it's on the list. If they
+have no list, they check the consensus parameter. If nether exist, then no
+FlashFlow deployment is allowed to measure this relay. Default: empty list.
```

Comment:
Are both of these parameters needed? Can't a non-empty FFAllowedCoordinators imply FFMeasurementsAllowed?

Reply:
FFAllowedCoordinators is also a consensus parameter. Consider the case when the consensus parameter is set but the operator doesn't want to be measured at all. Without the MeasurementsAllowed flag, the only way to effectively disallow measurement of yourself would be to set AllowedCoordinators to a list only a nonexistent coordinator TLS certificate fingerprint. That's hacky.

I've checked both with EntryNodes, a conceptually similar torrc option, and with dgoulet and arma. I've gathered that for these types of torrc options (LINELIST) Tor cannot tell the difference between an unspecified torrc option and a torrc option that specifies an empty list.